crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,550 @@
+        let required_set: HashSet<&String> = HashSet::from_iter(required.iter());
+        required_set
+            .difference(&self.names())
+            .map(|n| String::from(*n))
+            .collect()
+    }
+}
+
+pub(crate) type MaybeDevice = Result<Box<dyn ScmiDevice>, String>;
+type DeviceConstructor = fn(&DeviceProperties) -> MaybeDevice;
+
+/// Definition of a device kind.
+///
+/// Use [DeviceSpecification::new] to create it.
+pub(crate) struct DeviceSpecification {
+    /// Function to call to create the device.
+    ///
+    /// The device properties are those provided on the command line by the
+    /// user.
+    pub(crate) constructor: DeviceConstructor,
+    /// Short description of the device.
+    ///
+    /// Single line, not a complete sentence.
+    short_help: String,
+    /// Long description of the device.
+    ///
+    /// Complete sentences, can span multiple lines.
+    long_help: String,
+    /// Description of the device parameters available to the user.
+    ///
+    /// Each item in the vector corresponds to a single parameter description
+    /// and should start with the parameter name and a followup colon.
+    parameters_help: Vec<String>,
+}
+
+impl DeviceSpecification {
+    /// Creates a new device specification.
+    ///
+    /// See [DeviceSpecification] for the meaning of the arguments.
+    /// The device specification must be used in [available_devices] to
+    /// actually add the device.
+    fn new(
+        constructor: DeviceConstructor,
+        short_help: String,
+        long_help: String,
+        parameters_help: Vec<String>,
+    ) -> DeviceSpecification {
+        DeviceSpecification {
+            constructor,
+            short_help,
+            long_help,
+            parameters_help,
+        }
+    }
+}
+
+/// Mapping of device identifiers (names) to device specifications.
+///
+/// The string keys correspond to device identifiers specified on the command
+/// line.
+type NameDeviceMapping = HashMap<String, DeviceSpecification>;
+
+/// Creates device mapping and adds all the supported devices to it.
+///
+/// If you want to introduce a new kind of host device bindings, insert a
+/// device identifier + [DeviceSpecification] to [NameDeviceMapping] here.
+pub(crate) fn available_devices() -> NameDeviceMapping {
+    let mut devices: NameDeviceMapping = HashMap::new();
+    devices.insert(
+        String::from("fake"),
+        DeviceSpecification::new(
+            fake::FakeSensor::new_device,
+            String::from("fake accelerometer"),
+            String::from("A simple 3-axes sensor providing fake pre-defined values."),
+            vec![String::from(
+                "name: an optional name of the sensor, max. 15 characters",
+            )],
+        ),
+    );
+    devices.insert(
+        String::from("iio"),
+        DeviceSpecification::new(
+            iio::IIOSensor::new_device,
+            String::from("industrial I/O sensor"),
+            String::from(""),
```

Comment:
Shouldn't the description be optional then?

Reply:
Not sure, not having strong opinion either way.